### PR TITLE
feature(whl): add tabmwp env and prompt pg policy

### PR DESCRIPTION
**Pull request by kxzxvbk**
## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/model/template/nlp_pretrained_model.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import torch
```
Comment (PaParaZz1) [inline ding/model/template/nlp_pretrained_model.py:1]: rename to `language_transformers`

File: ding/model/template/nlp_pretrained_model.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+from ding.utils import MODEL_REGISTRY
+from torch import nn
+from transformers import AutoTokenizer, AutoModelForTokenClassification
```
Comment (PaParaZz1) [inline ding/model/template/nlp_pretrained_model.py:5]: add import try-except like [here](https://github.com/opendilab/DI-engine/blob/main/ding/utils/data/dataset.py#L51)

File: ding/model/template/nlp_pretrained_model.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        sentence_embedding = last_hidden_states[:, 0, :]  # len(input_list) x hidden_size
+        # print(f"sentence_embedding: {sentence_embedding}")
+
+        if self.linear:
+            sentence_embedding = self.linear(sentence_embedding)  # len(input_list) x embedding_size
+
+        return sentence_embedding
+
+    def forward(self, train_samples: list, candidate_samples: list) -> dict:
+        ctxt_embedding = self._calc_embedding(train_samples)
```
Comment (PaParaZz1) [inline ding/model/template/nlp_pretrained_model.py:51]: why namt `ctxt`

File: ding/model/template/tests/test_nlp_pretrained_model.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        cands_list = [problems[pid] for pid in cand_pids]
+
+        model = NLPPretrainedModel(model_name="bert-base-uncased", add_linear=True, embedding_size=256)
+        cands_embedding = model(cands_list)
+        assert cands_embedding.shape == (3, 256)
+        ctxt_embedding = model(ctxt_list)
+        assert ctxt_embedding.shape == (1, 256)
+
+        scores = torch.mm(ctxt_embedding, cands_embedding.t())
+        assert scores.shape == (1, 3)
```
Comment (PaParaZz1) [inline ding/model/template/tests/test_nlp_pretrained_model.py:27]: test free_encoder

File: dizoo/tabmwp/config/tabmwp_pg_config.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    ),
+    env_manager=dict(type='base'),
+    policy=dict(type='prompt_pg'),
+    replay_buffer=dict(type='naive'),
+)
+create_config = EasyDict(tabmwp_prompt_pg_config)
+
+if __name__ == '__main__':
+    from ding.entry import serial_pipeline_onpolicy
+    serial_pipeline_onpolicy((main_config, create_config), seed=0)
```
Comment (PaParaZz1) [inline dizoo/tabmwp/config/tabmwp_pg_config.py:63]: don't upload these json files, if necessary, you can upload these files to [here](https://opendilab.net/download/DI-zoo/)

File: dizoo/tabmwp/config/tabmwp_pg_config.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+            # (bool) Whether to normalize advantage. Default to False.
+            learning_rate=0.001,
+            # (float) loss weight of the value network, the weight of policy network is set to 1
+            entropy_weight=0.,
+            grad_norm=0.5,
+        ),
+        collect=dict(
+            # (int) collect n_sample data, train model 1 times
+            n_sample=20,
+            discount_factor=0.,
```
Comment (PaParaZz1) [inline dizoo/tabmwp/config/tabmwp_pg_config.py:43]: why `0.` here

File: dizoo/tabmwp/config/tabmwp_pg_config.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
+from easydict import EasyDict
+
+tabmwp_prompt_pg_config = dict(
+    exp_name='tabmwp_prompt_pg_seed0',
+    env=dict(
+        collector_env_num=1,
+        evaluator_env_num=1,
+        n_evaluator_episode=1,
+        stop_value=1,
+        cand_number=20,
```
Comment (PaParaZz1) [inline dizoo/tabmwp/config/tabmwp_pg_config.py:10]: move the arguments that are not frequently modified to env's `default_config`

File: dizoo/tabmwp/envs/test_tabmwp_env.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+            max_tokens=512,
+            top_p=1.,
+            frequency_penalty=0.,
+            presence_penalty=0.,
+            option_inds=["A", "B", "C", "D", "E", "F"],
+            api_key='',
+        )
+        config = EasyDict(config)
+        env = TabMWP(config)
+        env.seed(0)
```
Comment (PaParaZz1) [inline dizoo/tabmwp/envs/test_tabmwp_env.py:24]: why not test `reset` and `step`

File: ding/policy/prompt_pg.py [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+            output = self._model.forward(data['train_sample'], data['candidate_samples'])
+            # Generate actions.
+            act = []
+            mask = torch.zeros_like(output['logit'])
+            for ii in range(self._cfg.shot_number):
+                dist = torch.distributions.Categorical(logits=output['logit'] + mask)
+                actions = dist.sample()
+                act.append(actions)
+                for jj in range(actions.shape[0]):
+                    mask[jj][actions[jj]] = -1e30
```
Comment (PaParaZz1) [inline ding/policy/prompt_pg.py:166]: -1e9

File: ding/policy/prompt_pg.py [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+        pass
+
+    def _forward_eval(self, data: dict) -> dict:
+        data_id = list(data.keys())
+        data = default_collate(list(data.values()))
+        self._model.eval()
+        with torch.no_grad():
+            # Prepare train_sample (the question to be answered) and the candidate_samples (the prompts to be selected)
+            for ii in range(len(data['candidate_samples'])):
+                data['candidate_samples'][ii] = data['candidate_samples'][ii][0]
```
Comment (PaParaZz1) [inline ding/policy/prompt_pg.py:229]: can we execute this data transform in environment?

File: ding/policy/prompt_pg.py [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+                cand_samples[ii] = cand_samples[ii][0]
+            output = self._learn_model.forward(train_samples, cand_samples)
+            return_ = batch['return']
+            if self._cuda:
+                return_ = return_.to(self._device)
+
+            # calculate PG loss
+            real_act = []
+            for b in range(batch['action'].shape[0]):
+                tmp_act = []
```
Comment (PaParaZz1) [inline ding/policy/prompt_pg.py:101]: can we move this action transform in env?

File: ding/policy/prompt_pg.py [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+                assert len(tmp_act) == self._cfg.shot_number
+                real_act.append(tmp_act)
+            real_act = torch.tensor(real_act, device=self._device)  # shape: (B, shot_number)
+            # Calculate loss.
+            total_loss = 0
+            for ii in range(self._cfg.shot_number):
+                log_prob = output['dist'].log_prob(real_act[:, ii])
+                policy_loss = -(log_prob * return_).mean()
+                entropy_loss = -self._cfg.learn.entropy_weight * output['dist'].entropy().mean()
+                total_loss += policy_loss + entropy_loss
```
Comment (PaParaZz1) [inline ding/policy/prompt_pg.py:121]: I think we should only add entropy loss once

File: ding/policy/prompt_pg.py [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+        self._model.eval()
+        with torch.no_grad():
+            # Prepare train_sample (the question to be answered) and the candidate_samples (the prompts to be selected)
+            for ii in range(len(data['candidate_samples'])):
+                data['candidate_samples'][ii] = data['candidate_samples'][ii][0]
+            output = self._model.forward(data['train_sample'], data['candidate_samples'])
+            # Generate actions.
+            act = []
+            mask = torch.zeros_like(output['logit'])
+            for ii in range(self._cfg.shot_number):
```
Comment (PaParaZz1) [inline ding/policy/prompt_pg.py:234]: maybe we can abstarct the operation of the shot_number into a new head in `ding/model/common`

File: dizoo/tabmwp/envs/tabmwp_env.py [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+
+        # normalize the number in the text
+        prediction_norm = normalize_answer(prediction, self.problems[pid]['unit'])
+
+        if prediction_norm.lower() == normalize_answer(self.problems[pid]['answer'],
+                                                       self.problems[pid]['unit']).lower():
+            _reward = 1
+            self.correct_num += 1
+        else:
+            _reward = -1
```
Comment (PaParaZz1) [inline dizoo/tabmwp/envs/tabmwp_env.py:108]: why add this underline

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: dizoo/tabmwp/envs/tabmwp_env.py [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+        self.reward_space = gym.spaces.Box(
+            low=-1, high=1, shape=(1,), dtype=np.float32
+        )
+        self.correct_num = 0
+        assert self._args.engine in ['text-davinci-002', 'glm-10B', 'rwkv-7B', 'internlm-7B']
+        if self._args.engine == 'glm-10B' and TabMWP.model is None:
+            from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
+            TabMWP.tokenizer = AutoTokenizer.from_pretrained("THUDM/glm-10b", trust_remote_code=True)
+            model = AutoModelForSeq2SeqLM.from_pretrained("THUDM/glm-10b", trust_remote_code=True)
+            TabMWP.model = model.half().cuda()
```
Comment (PaParaZz1) [inline dizoo/tabmwp/envs/tabmwp_env.py:32]: move `model.cuda()` operation to `reset` method

File: dizoo/tabmwp/envs/tabmwp_env.py [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+from dizoo.tabmwp.envs.utils import *
+
+
+@ENV_REGISTRY.register('tabmwp')
+class TabMWP(BaseEnv):
+    model = None
+    tokenizer = None
+
+    def __init__(self, cfg):
+        self._args = cfg
```
Comment (PaParaZz1) [inline dizoo/tabmwp/envs/tabmwp_env.py:14]: rename `_args` to `cfg`

File: dizoo/tabmwp/config/tabmwp_pg_config.py [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+        cand_number=16,
+        train_number=20,
+        engine='text-davinci-002',
+        temperature=0.,
+        max_tokens=512,
+        top_p=1.,
+        frequency_penalty=0.,
+        presence_penalty=0.,
+        option_inds=["A", "B", "C", "D", "E", "F"],
+        api_key='',
```
Comment (PaParaZz1) [inline dizoo/tabmwp/config/tabmwp_pg_config.py:19]: add comments about this

File: ding/policy/prompt_pg.py [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+from ding.utils import POLICY_REGISTRY
+from ding.utils.data import default_collate, default_decollate
+from .base_policy import Policy
+
+
+@POLICY_REGISTRY.register('prompt_pg')
+class PromptPGPolicy(Policy):
+    r"""
+    Overview:
+        Policy class of Prompt Policy Gradient (PromptPG) algorithm.
```
Comment (PaParaZz1) [inline ding/policy/prompt_pg.py:16]: add paper link

File: ding/policy/prompt_pg.py [hunk 19]
```diff
@@ -0,0 +1,258 @@
...
+        return_infos = []
+        for i in range(0, len(data), self._cfg.learn.batch_size):
+            batch = default_collate(data[i:i + self._cfg.learn.batch_size])
+            # Prepare train_sample (the question to be answered) and the candidate_samples (the prompts to be selected)
+            train_samples, cand_samples = batch["obs"]["train_sample"], batch["obs"]["candidate_samples"]
+            for ii in range(len(cand_samples)):
+                cand_samples[ii] = cand_samples[ii][0]
+            output = self._learn_model.forward(train_samples, cand_samples)
+            return_ = batch['return']
+            if self._cuda:
```
Comment (PaParaZz1) [inline ding/policy/prompt_pg.py:95]: why cuda here, all the data has been moved to cuda in line 84

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/model/wrapper/model_wrappers.py [hunk 20]
```diff
@@ -411,6 +411,64 @@ def forward(self, *args, **kwargs):
...
+        output = self._model.forward(*args, **kwargs)
+        # Generate actions.
+        act = []
+        mask = torch.zeros_like(output['logit'])
+        for ii in range(shot_number):
+            dist = torch.distributions.Categorical(logits=output['logit'] + mask)
+            actions = dist.sample()
+            act.append(actions)
+            for jj in range(actions.shape[0]):
+                mask[jj][actions[jj]] = -1e30
```
Comment (PaParaZz1) [inline ding/model/wrapper/model_wrappers.py:459]: -1e8

File: ding/model/wrapper/model_wrappers.py [hunk 21]
```diff
@@ -411,6 +411,64 @@ def forward(self, *args, **kwargs):
...
+        mask = torch.zeros_like(output['logit'])
+        for ii in range(shot_number):
+            masked_logit = output['logit'] + mask
+            actions = masked_logit.argmax(dim=-1)
+            act.append(actions)
+            for jj in range(actions.shape[0]):
+                mask[jj][actions[jj]] = -1e8
+        # `act` is shaped (shot_num, B)
+        real_act = []
+        for b in range(act[0].shape[0]):
```
Comment (PaParaZz1) [inline ding/model/wrapper/model_wrappers.py:433]: add some comments for this transformaton

File: ding/model/wrapper/test_model_wrappers.py [hunk 22]
```diff
@@ -549,3 +549,15 @@ def test_transformer_memory_wrapper(self):
...
+        model = model_wrap(ActorMLP(), wrapper_name='combination_argmax_sample')
+        data = {'obs': torch.randn(4, 3)}
+        output = model.forward(shot_number=2, inputs=data)
+        assert output['action'].shape == (4, )
+
+    def test_combination_multinomial_sample_wrapper(self):
+        model = model_wrap(ActorMLP(), wrapper_name='combination_multinomial_sample')
+        data = {'obs': torch.randn(4, 3)}
+        output = model.forward(shot_number=2, inputs=data)
+        assert output['action'].shape == (4, )
```
Comment (PaParaZz1) [inline ding/model/wrapper/test_model_wrappers.py:563]: assert action range

File: dizoo/tabmwp/envs/tabmwp_env.py [hunk 23]
```diff
@@ -0,0 +1,241 @@
+import gym
+
+from ding.utils import ENV_REGISTRY
+from ding.envs import BaseEnv, BaseEnvTimestep
+from dizoo.tabmwp.envs.utils import *
```
Comment (PaParaZz1) [inline dizoo/tabmwp/envs/tabmwp_env.py:5]: add concrete import names

File: dizoo/tabmwp/envs/tabmwp_env.py [hunk 24]
```diff
@@ -0,0 +1,241 @@
...
+        self.action_space = gym.spaces.Discrete(self.cfg.cand_number * (self.cfg.cand_number - 1))
+        self.reward_space = gym.spaces.Box(
+            low=-1, high=1, shape=(1,), dtype=np.float32
+        )
+        self.correct_num = 0
+
+        # Initialize language model if needed.
+        assert self.cfg.engine in ['text-davinci-002', 'glm-10B', 'rwkv-7B', 'internlm-7B']
+        if self.cfg.engine == 'glm-10B' and TabMWP.model is None:
+            from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
```
Comment (PaParaZz1) [inline dizoo/tabmwp/envs/tabmwp_env.py:31]: add import check

File: dizoo/tabmwp/envs/tabmwp_env.py [hunk 25]
```diff
@@ -0,0 +1,241 @@
+import gym
```
Comment (PaParaZz1) [inline dizoo/tabmwp/envs/tabmwp_env.py:1]: add this env and algo in readme.md

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Review by PaParaZz1 — APPROVED** [review]

**Review by PaParaZz1 — APPROVED** [review]
